src/box/memtx_read_set.c

```diff
@@ -0,0 +1,225 @@
+
+int
+memtx_read_interval_cmpl(const struct memtx_read_interval *a,
+		      const struct memtx_read_interval *b)
+{
+	assert(a->index == b->index);
+	assert(a->left.key != NULL);
+	assert(a->right.key != NULL);
+	assert(b->left.key != NULL);
+	assert(b->right.key != NULL);
+
+	struct key_def *cmp_def = a->index->def->cmp_def;
+	int cmp = key_compare_ext(a->left.key, a->left.hint,
+				  b->left.key, b->left.hint, cmp_def);
+	if (cmp != 0)
+		return cmp;
+	if (a->left_belongs && !b->left_belongs)
+		return -1;
+	if (!a->left_belongs && b->left_belongs)
+		return 1;
+	if (is_inf(a->left.key) || is_inf(b->left.key))
+		return 0;
+
+	const char* a_key = a->left.key;
+	uint32_t a_parts = mp_decode_array(&a_key);
+	const char* b_key = b->left.key;
+	uint32_t b_parts = mp_decode_array(&b_key);
+	if (a->left_belongs)
+		return a_parts < b_parts ? -1 : a_parts > b_parts;
+	else
+		return a_parts > b_parts ? -1 : a_parts < b_parts;
+}
+
+int
+memtx_read_interval_cmpr(const struct memtx_read_interval *a,
+		      const struct memtx_read_interval *b)
+{
+	assert(a->index == b->index);
+	assert(a->left.key != NULL);
+	assert(a->right.key != NULL);
+	assert(b->left.key != NULL);
+	assert(b->right.key != NULL);
+
+	struct key_def *cmp_def = a->index->def->cmp_def;
+	int cmp = key_compare_ext(a->right.key, a->right.hint,
+				  b->right.key, b->right.hint, cmp_def);
+	if (cmp != 0)
+		return cmp;
+	if (a->right_belongs && !b->right_belongs)
+		return 1;
+	if (!a->right_belongs && b->right_belongs)
+		return -1;
+	if (is_inf(a->right.key) || is_inf(b->right.key))
+		return 0;
+
+	const char* a_key = a->right.key;
+	uint32_t a_parts = mp_decode_array(&a_key);
+	const char* b_key = b->right.key;
+	uint32_t b_parts = mp_decode_array(&b_key);
+	if (a->right_belongs)
+		return a_parts > b_parts ? -1 : a_parts < b_parts;
+	else
+		return a_parts < b_parts ? -1 : a_parts > b_parts;
+}
+
+bool
+memtx_read_interval_should_merge(const struct memtx_read_interval *l,
+			      const struct memtx_read_interval *r)
+{
+	assert(l->index == r->index);
+	assert(l->left.key != NULL);
+	assert(l->right.key != NULL);
+	assert(r->left.key != NULL);
+	assert(r->right.key != NULL);
+	assert(memtx_read_interval_cmpl(l, r) <= 0);
+
+	struct key_def *cmp_def = l->index->def->cmp_def;
+	int cmp = key_compare_ext(l->right.key, l->right.hint,
+				  r->left.key, r->left.hint, cmp_def);
+	if (cmp > 0)
+		return true;
+	if (cmp < 0)
+		return false;
+	if (l->right_belongs && r->left_belongs)
+		return true;
+	if (!l->right_belongs && !r->left_belongs)
+		return false;
+	if (is_inf(l->right.key) || is_inf(r->left.key))
+		return true;
+
+	const char* left_key = l->right.key;
+	uint32_t l_parts = mp_decode_array(&left_key);
+	const char* right_key = r->left.key;
+	uint32_t r_parts = mp_decode_array(&right_key);
+	if (l->right_belongs)
+		return l_parts <= r_parts;
+	else
+		return l_parts >= r_parts;
+}
+
+int
+memtx_entry_compare(struct memtx_entry left, struct memtx_entry right,
```

Comment:
Here and everywhere, struct memtx_entry is not big, but i think it's better to pass it to function by pointer.

Reply:
For what purpose? If it's not big, it's not a big deal, and when I see a function that needs passing by reference, my first thought is that this function needs to change something inside the struct, which is not true for this case.